File: engine/player.ts
import {GameHost, type HostFiles} from '@dreamfactory/engine/web/host';
import {installRestoredSaving} from './restoration/neutral-save';
import {parseSave} from './restoration/save-format';
import type {AudioSink, AudioChannel, PlayOpts, PlayHandle} from '@dreamfactory/engine/runtime/audio';
import type {DecodedAudio} from '@dreamfactory/engine/df/audio';
import {ALL_CHANNELS,resetScreenGamma,stepScreenGamma} from '@dreamfactory/engine/web/screen-gamma';

declare const __native:(method:string,args:string,buffer?:ArrayBuffer)=>string|ArrayBuffer|null;
const call=(method:string,args:any={})=>JSON.parse(__native(method,JSON.stringify(args)) as string || 'null');
const read=(path:string):Uint8Array|null=>{const b=__native('read',JSON.stringify({path}));return b instanceof ArrayBuffer?new Uint8Array(b):null;};
const write=(path:string,bytes:Uint8Array)=>{const r=JSON.parse(__native('write',JSON.stringify({path}),bytes.slice().buffer) as string);if(r?.error)throw Error(r.error);};
const events:any[]=[];
const emit=(type:string,data:any={})=>events.push({type,...data});
const fail=(e:any)=>{emit('log',{text:String(e)+(e?.stack?'\n'+e.stack:'')});emit('error',{text:String(e?.message||e)});};
export class Files implements HostFiles {
 disc:1|2=1; cache=new Map<string,Uint8Array>();
 constructor(readonly index:Record<string,string>){}
 key(name:string){return `${this.disc}/${name.toLowerCase().replace(/^.*[\\/:]/,'')}`;}
 provide=(name:string)=>{const key=this.key(name);if(this.cache.has(key))return this.cache.get(key)!;const p=this.index[key]||this.index[`${this.disc===1?2:1}/${key.slice(2)}`];const bytes=p?read(p):null;if(bytes)this.cache.set(key,bytes);return bytes;};
 async load(name:string){return this.provide(name);}
 setDisc(d:1|2){this.disc=d;}
 activeDisc(){return this.disc;}
 textEncoding(){return 'macintosh' as const;}
 has(name:string){return this.cache.has(this.key(name));}
 evict(name:string){const key=this.key(name),n=this.cache.get(key)?.length||0;this.cache.delete(key);return n;}
}
let nextAudio=0;
const audioBuffers=new Map<number,ArrayBuffer>();
class Audio implements AudioSink {
 entries=new Map<number,{done:boolean,channel:AudioChannel,overlap:boolean}>();
 play(channel:AudioChannel,audio:DecodedAudio,opts:PlayOpts={}):PlayHandle {
  if(!opts.overlap)this.halt(channel);
  const id=++nextAudio,entry={done:false,channel,overlap:!!opts.overlap};this.entries.set(id,entry);
  const pan=Math.max(-1,Math.min(1,opts.pan||0));
  const pcm=new ArrayBuffer(audio.samples.length*4),dv=new DataView(pcm);
  const gain=Math.max(0,Math.min(1,opts.volume??1));
  const l=gain*Math.cos((pan+1)*Math.PI/4),r=gain*Math.sin((pan+1)*Math.PI/4);
  for(let i=0;i<audio.samples.length;i++){const s=Math.max(-1,Math.min(1,audio.samples[i]))*32767;dv.setInt16(i*4,Math.round(s*l),true);dv.setInt16(i*4+2,Math.round(s*r),true);}
  audioBuffers.set(id,pcm);emit('audio_play',{id,channel,rate:audio.sampleRate,loop:!!opts.loop,samples:audio.samples.length});
  return {get done(){return entry.done;},stop:()=>{entry.done=true;this.entries.delete(id);emit('audio_stop',{id});}};
 }
 halt(channel:AudioChannel){for(const [id,e]of this.entries)if(e.channel===channel&&!e.overlap){e.done=true;this.entries.delete(id);emit('audio_stop',{id});}}
 isDone(channel:AudioChannel){return ![...this.entries.values()].some(e=>e.channel===channel&&!e.overlap&&!e.done);}
 setChannelVolume(channel:AudioChannel,volume:number){emit('audio_volume',{channel,volume});}
 setSuspended(on:boolean){emit('audio_pause',{on});}
 finish(id:number){const e=this.entries.get(id);if(e)e.done=true;this.entries.delete(id);}
}
const audio=new Audio();
const context:any={canvas:{width:512,height:384},font:'12px Arial',fillStyle:'#fff',strokeStyle:'#fff',lineWidth:1,commands:[],stack:[],version:0,
 putImageData(){this.commands=[];this.version++;},
 save(){this.stack.push([this.font,this.fillStyle,this.strokeStyle,this.lineWidth]);},
 restore(){[this.font,this.fillStyle,this.strokeStyle,this.lineWidth]=this.stack.pop();},
 fillText(text:string,x:number,y:number){this.commands.push({op:'text',text,x,y,font:this.font,color:this.fillStyle});},
 measureText(text:string){return {width:call('measure',{text,font:this.font})};},
 fillRect(x:number,y:number,w:number,h:number){this.commands.push({op:'rect',x,y,w,h,color:this.fillStyle});},
 strokeRect(x:number,y:number,w:number,h:number){this.commands.push({op:'stroke',x,y,w,h,color:this.strokeStyle,line:this.lineWidth});}
};
let host:GameHost, ready=false, pauseOwned=false, paused=false, busy=false, renderVersion=0;
let profiling=false,profileTick=0,profileRender=0;
let frameWaiters:(()=>void)[]=[];
let nextDialog=0;
const dialogs=new Map<number,(v:any)=>void>();
async function dialog(kind:string,text:string,value=''):Promise<any>{const id=++nextDialog;emit('dialog',{id,kind,text,value});return new Promise(resolve=>dialogs.set(id,resolve));}
async function frozen<T>(fn:()=>Promise<T>){const owns=!host.session.frozen;if(owns)host.session.freezeTime();try{return await fn();}finally{if(owns)host.session.thawTime();}}
async function save(bytes:Uint8Array){
 const name=await dialog('save','Name this saved game.',`${host.session.currentSetFile||'Voyage'} ${new Date().toISOString().slice(0,19).replace(/[:T]/g,'-')}`);
 if(!name)return;parseSave(bytes);write('save:'+name,bytes);emit('saved',{name});
}
async function chooseSave():Promise<Uint8Array|null>{
 const path=await dialog('load','Choose a saved game.');if(!path)return null;
 const bytes=read(path);if(!bytes)throw Error('Unable to read this save.');parseSave(bytes);return bytes;
}
async function boot(config:any){
 const files=new Files(config.index);
 host=new GameHost(files,audio,{log:text=>emit('log',{text}),hud:text=>emit('status',{text}),showStage:()=>emit('stage')});
 if(config.testing)(globalThis as any).titanicTesting={host,read,parseSave};
 const s=host.session;s.pictureMode='sharp';s.hasRealFrames=true;
 s.nextFrame=()=>new Promise(resolve=>frameWaiters.push(resolve));
 installRestoredSaving(s);
 s.onNoteDialog=async text=>{await frozen(()=>dialog('note',text));};
 s.onQuestionDialog=async text=>!!(await frozen(()=>dialog('question',text)));
 s.onTextDialog=async(text,initial)=>(await frozen(()=>dialog('text',text,initial)))||'';
 s.onQuit=()=>emit('quit');
 s.onSaveGame=async bytes=>{try{await save(bytes);}catch(e){fail(e);}};
 s.onLoadGame=async()=>{try{return await chooseSave();}catch(e){fail(e);return null;}};
 host.director.onCursor=name=>emit('cursor',{name});
 await host.preload();ready=true;emit('ready');
 if(config.save){const bytes=read(config.save);if(!bytes)throw Error('Cannot read save');parseSave(bytes);await host.loadSavedGame(bytes);}
 else await s.track(host.coldBoot(),'coldBoot');
}
function tick(dt:number){
 if(!host)return;
 const started=profiling?Date.now():0;
 if(!paused){(globalThis as any).__time+=dt;const now=(globalThis as any).__time;
 const timers=(globalThis as any).__timers;(globalThis as any).__timers=[];
 for(const t of timers)if(t.at<=now)t.fn();else(globalThis as any).__timers.push(t);
 host.director.tick(now);const waiters=frameWaiters;frameWaiters=[];for(const r of waiters)r();
 const renderStarted=profiling?Date.now():0;
 host.director.render(context);
 if(profiling){profileTick+=renderStarted-started;profileRender+=Date.now()-renderStarted;}
 }
}
async function command(c:any){
 if(c.action==='reply'){const resolve=dialogs.get(c.id);dialogs.delete(c.id);resolve?.(c.value);return;}
 if(c.action==='audio_done'){audio.finish(c.id);return;}
 if(!host)return;const s=host.session,d=host.director;
 if(c.action==='pause'){
  if(c.on===paused)return;paused=c.on;
  if(paused){pauseOwned=!s.frozen;if(pauseOwned)s.freezeTime();s.pointerDown=false;d.release(-1,-1);}
  else {if(pauseOwned)s.thawTime();pauseOwned=false;}return;
 }
 if(!ready||paused)return;
 if(c.action==='pointer'){s.shiftDown=!!c.shift;s.setPointer(c.x,c.y);
  if(c.kind==='press'){s.pointerDown=true;await s.track(d.press(c.x,c.y),'pointer press');}
  else if(c.kind==='release'){s.pointerDown=false;d.release(c.x,c.y);}
  else if(!s.pointerDown)await d.hover(c.x,c.y);return;
 }
 if(c.action==='key'){
  const key=c.key;if(['uparrow','leftarrow','rightarrow'].includes(key)&&host.viewer&&(s.viewShowing||!s.stageCtrl.keydownTarget()))await s.track(host.viewer.pressNav(key),'navigation');
  else await s.track(d.keyDown(key,!!c.special),'key');return;
 }
 if(c.action==='gamma'){if(c.key===9)resetScreenGamma();else stepScreenGamma(c.key%2===0,c.key<3?ALL_CHANNELS:[c.key<5,c.key>=5&&c.key<7,c.key>=7]);return;}
 if(busy)return;busy=true;
 try{
 if(c.action==='save'){
  if(!s.currentSetFile||d.inputLocked||!s.viewShowing)throw Error('Finish the conversation or animation, then save while exploring.');
  await frozen(async()=>{const bytes=s.snapshotSave();if(!bytes)throw Error('Unable to save');await save(bytes);});
 }
 if(c.action==='load'||c.action==='import')await frozen(async()=>{
  const path=c.path||await dialog(c.action==='import'?'import':'load','Choose a saved game.');if(!path)return;
  const bytes=read(path);if(!bytes)throw Error('Cannot read saved game.');parseSave(bytes);
  // Restart the VM only after validation. This also abandons any old script continuations.
  if(c.action==='import')write('save:'+path.replace(/\\/g,'/').split('/').pop(),bytes);
  emit('restart',{save:path});
 });
 if(c.action==='new')await frozen(async()=>{if(await dialog('question','Return to the main menu? Unsaved progress will be lost.'))emit('restart');});
 }finally{busy=false;}
}
Object.assign(globalThis,{
 titanicProfile:(enabled:boolean)=>{profiling=enabled;},
 titanicTimings:()=>{const result=JSON.stringify({tick_ms:profileTick,render_ms:profileRender});profileTick=0;profileRender=0;return result;},
 titanicBoot:(c:any)=>void boot(c).catch(fail),
 titanicTick:tick,
 titanicCommand:(c:any)=>void command(c).catch(fail),
 titanicEvents:()=>JSON.stringify(events.splice(0)),
 titanicFrame:()=>{if(context.version===renderVersion)return null;renderVersion=context.version;return host.screen.frame.buffer;},
 titanicOverlay:()=>JSON.stringify(context.commands),
 titanicAudio:(id:number)=>{const b=audioBuffers.get(id);audioBuffers.delete(id);return b;},
 titanicState:()=>JSON.stringify(host?{ready,set:host.session.currentSetFile,scene:host.session.currentSceneName(),view:host.session.currentViewName(),disc:host.files.activeDisc?.(),movie:host.director.movieFile,choices:host.director.choices,regions:host.director.movieRegions,frame:host.session.frameCounter,paused,inputLocked:host.director.inputLocked,viewShowing:host.session.viewShowing,frozen:host.session.frozen}:{}),
});

File: internal/engine/player.go
package engine

import (
	"fmt"
	"math"
	"strings"
	"time"

	"github.com/itskenny0/titanic-godot/internal/hdpack"
	"github.com/itskenny0/titanic-godot/internal/save"
)

// PlayerBridge is the complete platform boundary. Godot owns file dialogs,
// storage and font metrics. Calls from scripts are marshalled back to Pump's
// thread; the exported C entry points must lock that thread for their duration.
type PlayerBridge interface {
	Read(string) ([]byte, error)
	Write(string, []byte) error
	Measure(string, string) (float64, error)
}
type PlayerConfig struct {
	DisableAutosave bool              `json:"disable_autosave"`
	HDPack          string            `json:"hd_pack"`
	Index           map[string]string `json:"index"`
	Save            string            `json:"save"`
	Testing         bool              `json:"testing"`
}
type PlayerCommand struct {
	Action             string `json:"action"`
	Kind               string `json:"kind"`
	Key                any    `json:"key"`
	X, Y               float64
	Radius             float64 `json:"radius"`
	Shift, Special, On bool
	ID                 uint64 `json:"id"`
	Value              any    `json:"value"`
	Path               string `json:"path"`
}
type playerDialog struct {
	ready bool
	value any
}
type Player struct {
	Bridge                                               PlayerBridge
	Host                                                 *GameHost
	Audio                                                *HostAudio
	Context                                              *DrawContext
	Ready, Paused                                        bool
	Testing                                              bool
	Now                                                  func() time.Time
	pauseOwned, busy                                     bool
	pointerOffsetX, pointerOffsetY                       float64
	now                                                  float64
	renderVersion                                        uint64
	events                                               []any
	nextDialog                                           uint64
	dialogs                                              map[uint64]*playerDialog
	testFailTick                                         bool
	profiling                                            bool
	profileTick, profileRender                           time.Duration
	autosaveDisabled, checkpointArmed, checkpointPending bool
	checkpointQuiet                                      float64
}

func NewPlayer(bridge PlayerBridge) *Player {
	p := &Player{Bridge: bridge, Audio: new(HostAudio), Context: NewDrawContext(ScreenWidth, ScreenHeight), Now: time.Now, dialogs: map[uint64]*playerDialog{}}
	p.Context.Measure = func(text, font string) float64 {
		v, err := p.effect(func() (any, error) { return p.Bridge.Measure(text, font) })
		if err != nil {
			p.fail(err)
			return 0
		}
		return v.(float64)
	}
	return p
}
func (p *Player) effect(run func() (any, error)) (any, error) {
	if p.Host != nil {
		if task := p.Host.Session.Executor.Current(); task != nil {
			return task.Effect(run)
		}
	}
	return run()
}
func (p *Player) read(path string) ([]byte, error) {
	v, err := p.effect(func() (any, error) { return p.Bridge.Read(path) })
	if err != nil {
		return nil, err
	}
	if v == nil {
		return nil, nil
	}
	return v.([]byte), nil
}
func (p *Player) write(path string, data []byte) error {
	_, err := p.effect(func() (any, error) { return nil, p.Bridge.Write(path, data) })
	return err
}
func (p *Player) flushAudio() {
	for _, event := range p.Audio.DrainEvents() {
		p.events = append(p.events, event)
	}
}
func (p *Player) emit(kind string, fields map[string]any) {
	p.flushAudio()
	if fields == nil {
		fields = map[string]any{}
	}
	fields["type"] = kind
	p.events = append(p.events, fields)
}
func (p *Player) fail(err error) {
	if err != nil {
		p.emit("log", map[string]any{"text": err.Error()})
		p.emit("error", map[string]any{"text": err.Error()})
	}
}
func (p *Player) Boot(config PlayerConfig) error {
	if p.Host != nil {
		return fmt.Errorf("player already booted")
	}
	p.Testing = config.Testing
	p.autosaveDisabled = config.DisableAutosave
	files := NewFiles(config.Index, p.read)
	p.Host = NewGameHost(files, p.Audio, HostUI{Log: func(text string) { p.emit("log", map[string]any{"text": text}) }, HUD: func(text string) { p.emit("status", map[string]any{"text": text}) }, ShowStage: func() { p.emit("stage", nil) }})
	if config.HDPack != "" {
		pack, err := hdpack.Open(config.HDPack, p.read, func(text string) { p.emit("log", map[string]any{"text": text}) })
		if err != nil {
			p.emit("log", map[string]any{"text": fmt.Sprintf("HD pack unavailable: %v; using original artwork", err)})
		} else {
			p.Host.Screen().HD = NewHDSurface(pack, ScreenWidth, ScreenHeight)
			p.emit("log", map[string]any{"text": fmt.Sprintf("HD pack: %d images, 2x display, 24 MiB image cache", len(pack.Manifest.Images))})
		}
	}
	s := p.Host.Session
	s.PictureMode = "sharp"
	s.HasRealFrames = true
	s.OnNoteDialog = func(text string) error { return p.frozen(func() error { p.dialog("note", text, ""); return nil }) }
	s.OnQuestionDialog = func(text string) (bool, error) {
		var yes bool
		err := p.frozen(func() error { yes = playerTruthy(p.dialog("question", text, "")); return nil })
		return yes, err
	}
	s.OnTextDialog = func(text, initial string) (string, error) {
		var value string
		err := p.frozen(func() error { value = playerString(p.dialog("text", text, initial)); return nil })
		return value, err
	}
	s.OnQuit = func() error { p.emit("quit", nil); return nil }
	s.OnSaveGame = func(data []byte, _ string) error { p.fail(p.save(data)); return nil }
	s.OnLoadGame = func(string) ([]byte, error) { data, err := p.chooseSave(); p.fail(err); return data, nil }
	p.Host.Director.OnCursor = func(name string) { p.emit("cursor", map[string]any{"name": name}) }
	p.Host.Director.Movies.OnStarted = p.checkpointMovieStarted
	p.Host.Director.Movies.OnCutsceneFinished = p.checkpointMovieFinished
	s.Track("coldBoot", false, func(*Task) error {
		p.Host.Preload()
		p.Ready = true
		p.emit("ready", nil)
		if config.Save != "" {
			data, err := p.read(config.Save)
			if err != nil {
				return err
			}
			if data == nil {
				return fmt.Errorf("cannot read save")
			}
			if _, err := save.Parse(data); err != nil {
				return err
			}
			return p.Host.LoadSavedGame(data)
		}
		return p.Host.ColdBoot()
	})
	p.pump(false)
	return nil
}
func (p *Player) Close() {
	if p.Host != nil {
		p.Host.Close()
		p.Host = nil
	}
	clear(p.dialogs)
}
func (p *Player) pump(frame bool) {
	if p.Host == nil {
		return
	}
	s := p.Host.Session
	for _, done := range s.Pump(s.Executor.Now(), frame, 10000) {
		p.fail(done.Err)
	}
}
func (p *Player) Tick(dt float64) error {
	if p.testFailTick {
		p.testFailTick = false
		return fmt.Errorf("expected tick failure test")
	}
	if p.Host == nil || p.Paused {
		return nil
	}
	if math.IsNaN(dt) || math.IsInf(dt, 0) || dt < 0 {
		return fmt.Errorf("invalid frame duration")
	}
	began := time.Now()
	p.now += dt
	if _, err := p.Host.Director.Tick(p.now); err != nil {
		return err
	}
	p.pump(true)
	renderStart := time.Now()
	if err := p.Host.Director.Render(p.Context); err != nil {
		return err
	}
	p.serviceCheckpoint(dt)
	if p.profiling {
		p.profileTick += renderStart.Sub(began)
		p.profileRender += time.Since(renderStart)
	}
	return nil
}
func (p *Player) dialog(kind, text, value string) any {
	task := p.Host.Session.Executor.Current()
	if task == nil {
		panic("dialog outside game task")
	}
	p.nextDialog++
	id := p.nextDialog
	d := &playerDialog{}
	p.dialogs[id] = d
	defer delete(p.dialogs, id)
	p.emit("dialog", map[string]any{"id": id, "kind": kind, "text": text, "value": value})
	task.Wait(func() bool { return d.ready })
	return d.value
}
func (p *Player) frozen(run func() error) error {
	clock := p.Host.Session.Clock
	owns := !clock.Frozen()
	if owns {
		clock.Freeze()
		defer clock.Thaw()
	}
	return run()
}
func (p *Player) save(data []byte) error {
	label := p.Host.Session.CurrentSetFile
	if label == "" {
		label = "Voyage"
	}
	name := playerString(p.dialog("save", "Name this saved game.", label+" "+p.Now().UTC().Format("2006-01-02-15-04-05")))
	if name == "" {
		return nil
	}
	if _, err := save.Parse(data); err != nil {
		return err
	}
	if err := p.write("save:"+name, data); err != nil {
		return err
	}
	p.emit("saved", map[string]any{"name": name})
	return nil
}
func (p *Player) chooseSave() ([]byte, error) {
	path := playerString(p.dialog("load", "Choose a saved game.", ""))
	if path == "" {
		return nil, nil
	}
	data, err := p.read(path)
	if err != nil {
		return nil, err
	}
	if data == nil {
		return nil, fmt.Errorf("unable to read this save")
	}
	if _, err := save.Parse(data); err != nil {
		return nil, err
	}
	return data, nil
}
func playerString(value any) string {
	if !playerTruthy(value) {
		return ""
	}
	if s, ok := value.(string); ok {
		return s
	}
	return fmt.Sprint(value)
}
func playerTruthy(value any) bool {
	switch v := value.(type) {
	case nil:
		return false
	case bool:
		return v
	case string:
		return v != ""
	case float64:
		return v != 0 && !math.IsNaN(v)
	case int:
		return v != 0
	}
	return true
}
func (p *Player) Command(c PlayerCommand) error {
	if c.Action == "reply" {
		if d := p.dialogs[c.ID]; d != nil {
			d.value, d.ready = c.Value, true
		}
		if !p.Paused {
			p.pump(false)
		}
		return nil
	}
	if c.Action == "audio_done" {
		p.Audio.Finish(c.ID)
		if !p.Paused {
			p.pump(false)
		}
		return nil
	}
	if p.Host == nil {
		return nil
	}
	s, d := p.Host.Session, p.Host.Director
	if c.Action == "autosave_enabled" {
		p.autosaveDisabled = !c.On
		p.checkpointPending, p.checkpointQuiet = false, 0
		return nil
	}
	if c.Action == "pause" {
		if c.On == p.Paused {
			return nil
		}
		p.Paused = c.On
		if p.Paused {
			p.pauseOwned = !s.Clock.Frozen()
			if p.pauseOwned {
				s.Clock.Freeze()
			}
			s.PointerDown = false
			p.pointerOffsetX, p.pointerOffsetY = 0, 0
			d.Release(-1, -1)
		} else {
			if p.pauseOwned {
				s.Clock.Thaw()
			}
			p.pauseOwned = false
		}
		return nil
	}
	if !p.Ready || p.Paused {
		return nil
	}
	switch c.Action {
	case "pointer":
		if c.Kind == "press" {
			x, y := p.assistedPoint(c.X, c.Y, c.Radius)
			p.pointerOffsetX, p.pointerOffsetY = x-c.X, y-c.Y
		}
		if c.Kind == "press" || s.PointerDown {
			c.X += p.pointerOffsetX
			c.Y += p.pointerOffsetY
		}
		if c.Kind == "release" {
			p.pointerOffsetX, p.pointerOffsetY = 0, 0
		}
		s.ShiftDown = c.Shift
		s.SetPointer(c.X, c.Y)
		if c.Kind == "press" {
			s.PointerDown = true
			if err := d.Press(c.X, c.Y); err != nil {
				return err
			}
		} else if c.Kind == "release" {
			s.PointerDown = false
			p.pointerOffsetX, p.pointerOffsetY = 0, 0
			d.Release(c.X, c.Y)
		} else if !s.PointerDown {
			s.Track("hover", true, func(*Task) error { _, err := d.Hover(c.X, c.Y); return err })
		}
	case "key":
		key := playerString(c.Key)
		s.Track("key:"+key, false, func(*Task) error {
			if (key == "uparrow" || key == "leftarrow" || key == "rightarrow") && p.Host.Viewer != nil && (s.ViewShowing() || s.StageCtrl.KeydownTarget() == nil) {
				return p.Host.Viewer.PressNav(key)
			}
			_, err := d.KeyDown(key, c.Special)
			return err
		})
	case "gamma":
		key := 0
		switch n := c.Key.(type) {
		case float64:
			key = int(n)
		case int:
			key = n
		}
		if key == 9 {
			d.Gamma.Reset()
		} else {
			channels := AllGammaChannels
			if key >= 3 {
				channels = [3]bool{key < 5, key >= 5 && key < 7, key >= 7}
			}
			d.Gamma.Step(key%2 == 0, channels)
		}
	default:
		if p.busy {
			return nil
		}
		p.busy = true
		// Menu commands are outside the game's busy-script set, like the original
		// command layer. Saving must not reject its own request as an active script.
		s.Track("menu:"+c.Action, true, func(*Task) error { defer func() { p.busy = false }(); return p.menuCommand(c) })
	}
	p.pump(false)
	return nil
}
func (p *Player) menuCommand(c PlayerCommand) error {
	s := p.Host.Session
	switch c.Action {
	case "save":
		if !p.canSaveGame() {
			return fmt.Errorf("finish the conversation or animation, then save while exploring")
		}
		return p.frozen(func() error {
			data, err := s.SnapshotSave()
			if err != nil {
				return err
			}
			return p.save(data)
		})
	case "load", "import":
		return p.frozen(func() error {
			path := c.Path
			if path == "" {
				path = playerString(p.dialog(c.Action, "Choose a saved game.", ""))
			}
			if path == "" {
				return nil
			}
			data, err := p.read(path)
			if err != nil {
				return err
			}
			if data == nil {
				return fmt.Errorf("cannot read saved game")
			}
			if _, err := save.Parse(data); err != nil {
				return err
			}
			if c.Action == "import" {
				name := strings.ReplaceAll(path, "\\", "/")
				name = name[strings.LastIndex(name, "/")+1:]
				if err := p.write("save:"+name, data); err != nil {
					return err
				}
			}
			p.emit("restart", map[string]any{"save": path})
			return nil
		})
	case "new":
		return p.frozen(func() error {
			if playerTruthy(p.dialog("question", "Return to the main menu? Unsaved progress will be lost.", "")) {
				p.emit("restart", nil)
			}
			return nil
		})
	}
	return nil
}
func (p *Player) Events() []any {
	p.flushAudio()
	events := p.events
	p.events = nil
	if events == nil {
		return []any{}
	}
	return events
}
func (p *Player) Frame() []byte {
	if p.Host == nil || p.Context.Version == p.renderVersion {
		return nil
	}
	p.renderVersion = p.Context.Version
	screen := p.Host.Screen()
	if screen.HD != nil && screen.HD.Valid && screen.HD.Hits > screen.HD.FrameStartHits {
		return screen.HD.Pixels
	}
	return screen.Frame
}
func (p *Player) Overlay() []DrawCommand {
	if p.Context.Commands == nil {
		return []DrawCommand{}
	}
	return p.Context.Commands
}
func (p *Player) TakeAudio(id uint64) []byte { return p.Audio.TakePCM(id) }
func (p *Player) Profile(on bool)            { p.profiling = on }
func (p *Player) Timings() map[string]float64 {
	out := map[string]float64{"tick_ms": float64(p.profileTick) / float64(time.Millisecond), "render_ms": float64(p.profileRender) / float64(time.Millisecond)}
	p.profileTick, p.profileRender = 0, 0
	return out
}
func (p *Player) State() map[string]any {
	if p.Host == nil {
		return map[string]any{}
	}
	s, d := p.Host.Session, p.Host.Director
	var movie any
	if d.Movies.Playing() {
		movie = d.Movies.PlayingFile()
	}
	regions := []map[string]any{}
	for _, r := range d.Movies.WaitingRegions() {
		regions = append(regions, map[string]any{"type": r.Type, "target": r.Target, "event": r.Event, "sound": r.Sound, "x0": r.X0, "y0": r.Y0, "x1": r.X1, "y1": r.Y1, "record": r.Record})
	}
	choices := []map[string]any{}
	for _, c := range d.Choices() {
		choices = append(choices, map[string]any{"text": c.Text, "id": c.ID})
	}
	hdHits := uint64(0)
	if d.Screen.HD != nil {
		hdHits = d.Screen.HD.Hits
	}
	return map[string]any{"hd_hits": hdHits, "ready": p.Ready, "set": s.CurrentSetFile, "scene": s.CurrentSceneName(), "view": s.CurrentViewName(), "disc": p.Host.Files.ActiveDisc(), "movie": movie, "choices": choices, "regions": regions, "frame": s.Clock.FrameCounter, "paused": p.Paused, "inputLocked": d.InputLocked(), "viewShowing": s.ViewShowing(), "frozen": s.Clock.Frozen()}
}
